src/Microsoft.AspNetCore.Proxy/ProxyMiddleware.cs

```diff
@@ -59,8 +63,51 @@ public ProxyMiddleware(RequestDelegate next, IOptions<ProxyOptions> options)
             _httpClient = new HttpClient(_options.BackChannelMessageHandler ?? new HttpClientHandler());
         }
 
-        public async Task Invoke(HttpContext context)
+        public Task Invoke(HttpContext context) => context.WebSockets.IsWebSocketRequest ? HandleWebSocketRequest(context) : HandleHttpRequest(context);
+
+        private async Task HandleWebSocketRequest(HttpContext context)
+        {
+            using (var client = new ClientWebSocket())
+            {
+                foreach (var kv in context.Request.Headers)
+                {
+                    if (!NotForwardedWebSocketHeaders.Contains(kv.Key))
+                    {
+                        client.Options.SetRequestHeader(kv.Key, kv.Value);
+                    }
+                }
+
+                var wsScheme = string.Equals(_options.Scheme, "https", StringComparison.OrdinalIgnoreCase) ? "wss" : "ws";
+                var uriString = $"{wsScheme}://{_options.Host}:{_options.Port}{context.Request.PathBase}{context.Request.Path}{context.Request.QueryString}";
+
+                await client.ConnectAsync(new Uri(uriString), CancellationToken.None);
+                using (var server = await context.WebSockets.AcceptWebSocketAsync(client.SubProtocol))
+                {
+                    await Task.WhenAll(PumpWebSocket(client, server), PumpWebSocket(server, client));
+                }
+            }
+        }
+
+        private static async Task PumpWebSocket(WebSocket source, WebSocket dest)
         {
+            var buffer = new byte[4096];
+            while (true)
+            {
+                var res = await source.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+                if (res.MessageType == WebSocketMessageType.Close)
+                {
+                    await dest.CloseAsync(source.CloseStatus.Value, source.CloseStatusDescription, CancellationToken.None);
```

Comment:
CloseOutputAsync only sends a close. CloseAsync sends a close and then starts a read loop waiting for a close response. Since you have a separate thread handling the communication the other direction, you only want CloseOutputAsync in each thread.